feat!: Add OpacityProvider

ufrshubham opened the pull request:
# Description

This PR adds a new `OpacityProvider` interface which should be implemented by components that want to use `OpacityEffect`. This makes `OpacityEffect` decoupled from `HasPaint`.


## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have followed the [Contributor Guide] when preparing my PR.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples` or `docs`.


## Breaking Change?
<!--
Would your PR require Flame users to update their apps following your change?

If yes, then the title of the PR should include "!" (for example, `feat!:`, `fix!:`). See
[Conventional Commit] for details. Also, for a breaking PR uncomment and fill in the "Migration
instructions" section below.
-->

- [x] Yes, this PR is a breaking change.
- [ ] No, this PR is not a breaking change.

### Migration instructions

`OpacityEffect`'s `paintId` string input is replaced by `target` property.
E.g.: `paintId: 'bluePaint'` should now be replaced with `target: component.opacityProviderForPaint('bluePaint')`

## Related Issues

Closes #2053 

<!-- Links -->
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Conventional Commit]: https://conventionalcommits.org

ufrshubham commented:
@st-pasha, this draft fixes the issue of using generic a single `paintId`. But for allowing `OpacityEffect` to operate on multiple `paints`, I think we will have to allow multiple targets for `OpacityEffect`. 

With single `target`, I am not sure what the `opacity` and `alpha` getters should return. It is quite possible that all paints do not have same opacity.

ufrshubham commented on packages/flame/lib/src/effects/provider_interfaces.dart:
```diff
@@ -50,3 +50,12 @@ abstract class SizeProvider {
+  set opacity(double value);
+
+  int get alpha;
+  set alpha(int value);
```
They are slightly different representation of the same idea. `alpha` is an integer in the range [0-255] and opacity is a double in the range [0-1]. I added both of them just to avoid making changes to `OpacityEffect`'s calculations as they use alpha right now.

st-pasha replied:
I've updated the logic of `OpacityEffect` to use opacities instead, in PR #2064.

ufrshubham replied:
Cool! I'll update mine once that PR is merged.

ufrshubham commented on packages/flame/lib/src/components/mixins/has_paint.dart:
```diff
@@ -97,4 +98,27 @@ mixin HasPaint<T extends Object> on Component {
+  double get opacity => paint.color.opacity;
+
+  @override
+  set opacity(double value) => paint.color = paint.color.withOpacity(value);
```
Done

st-pasha commented on packages/flame/lib/src/components/mixins/has_paint.dart:
```diff
@@ -97,4 +98,32 @@ mixin HasPaint<T extends Object> on Component {
+  double get opacity => target.getOpacity(paintId: paintId);
+
+  @override
+  set opacity(double value) => target.setOpacity(opacity, paintId: paintId);
```
```suggestion
  set opacity(double value) => target.setOpacity(value, paintId: paintId);
```
I guess it means we don't cover this provider tests?

ufrshubham replied:
Oh!! Good catch! I've updated this and a test to check this. I think we can create an issue for adding unit tests for `HasPaint` mixin.

st-pasha commented on packages/flame/lib/src/effects/opacity_effect.dart:
```diff
@@ -1,75 +1,75 @@
 
   double _opacityOffset;
   double _roundingError = 0.0;
-  final String? paintId;
```
Looks like this `paintId` was buggy, so good riddance. Generally, `HasPaint` is supposed to have a `<T>` generic parameter, so that `paintId` would be of type `T?`. But the class was declared as `ComponentEffect<HasPaint>`, thus `T=Object`. But when `T=Object` the paint ids are not allowed (there is an assert for that).

Not sure how (and whether) it all worked before, but good thing we don't have to worry about that anymore.

ufrshubham replied:
Yeah, I don't think most users know that `HasPaint` can actually hold multiple paints. Even our `Flower` class from examples stores multiple paints within itself, even though it has `HasPaint`  mixin 😅

st-pasha commented on packages/flame/lib/src/components/mixins/has_paint.dart:
```diff
@@ -97,4 +98,32 @@ mixin HasPaint<T extends Object> on Component {
+    }
+  }
+
+  OpacityProvider opacityProviderForPaint(T paintId) {
```
Maybe call it `opacityOfPaint(paintId)` for short?

`OpacityEffect(target: component.opacityOfPaint('bg'))` sounds nicer.

spydon replied:
It sounds like you'd get back the `double` from `opacityOfPaint` though?

st-pasha replied:
Well, you'd get something that has `.opacity` property; it's a bit like getting an `Opacity` object instead of a simple value. And there is definitely no confusion when used as a target of an `OpacityEffect`.

spydon replied:
There is no confusion when used as a target, but the user will get this method suggested when doing other things on the component too. Since we have no opacity object I don't think any user would expect that.

ufrshubham replied:
A short name would have been better, but just like @spydon said, `opacityOfPaint` will be a misnomer. How about `opacityProviderOf`? The input parameter already suggests that it operates on paint

st-pasha commented on packages/flame/lib/src/components/mixins/has_paint.dart:
```diff
@@ -97,4 +98,32 @@ mixin HasPaint<T extends Object> on Component {
+  OpacityProvider opacityProviderForPaint(T paintId) {
+    return _ProxyOpacityProvider(paintId, this);
+  }
+}
```
How about adding a provider for all opacities at once? This seems to be a common case, where you have a multi-paint component, and want to fade it out as a whole.

ufrshubham replied:
> This seems to be a common case, where you have a multi-paint component, and want to fade it out as a whole.

This use case is already handled by `opacity` setter of `HasPaint`. Even if we add a provider for all paints, how do we handle `opacity` getter if all paints do not have the same `opacity`?

st-pasha replied:
I would say that if you have a class with 2 paints, one with opacity 90% and another at 25%, then `get`ting the overall opacity should return 90% (the most opaque component), and `set`ting it to, say 50%`, should make the first paint go to 50% and the second paint to (25/90*50)% -- that is, the ratio of the opacities should remain the same.

This means the opacity provider has to store these ratios at the time of construction, and then use them later on.

spydon replied:
Isn't it overkill to have logic for that, the user can just handle the colors separately if they need a set ratio between the opacities?

It also gets a bit weird since the colors with less opacity never can get above what they started at if another color started on 1.0.

st-pasha replied:
> Isn't it overkill to have logic for that in here?

Not an overkill, since the logic will be in a separate `OpacityProvider` class.

> It also gets a bit weird since the colors with less opacity never can get above what they started at if another color started on 1.0.

Why would that be a problem? If you draw something with a faint color on top of a solid color -- for example, a watermark on a page -- then you'd want it to keep roughly the same opacity ratio relative to the whole component. On the other hand, if you wanted to make the water mark more or less visible -- then the OpacityEffect can be applied to that individual paint.

spydon replied:
> Why would that be a problem? If you draw something with a faint color on top of a solid color -- for example, a watermark on a page -- then you'd want it to keep roughly the same opacity ratio relative to the whole component. On the other hand, if you wanted to make the water mark more or less visible -- then the OpacityEffect can be applied to that individual paint.

Sure, but that is only one usecase, I'm sure there are others.

> then the OpacityEffect can be applied to that individual paint.

True, that would get a very strange behavior if it wasn't done in this manner, I'm convinced.

ufrshubham replied:
> for example, a watermark on a page -- then you'd want it to keep roughly the same opacity ratio relative to the whole component

So if such a component is completely faded out, both paint opacities will become zero. After this, if the component is faded in using a new `OpacityProvider`, it will not maintain the original ratios.

> `get`ting the overall opacity should return 90% (the most opaque component)

This approach doesn't really define an absolute true opacity. We are returning the opacity of most **opaque** paint and that can change anytime depending on how other opacity effects modify the paints. Even paints can be added and removed by the users anytime. I think this will involve a lot of book keeping to maintain the ratios.

st-pasha replied:
Surely there could be some corner cases when the effect doesn't work quite right. And if someone encounters such case in real life, they will have to find a way to deal with that problem (and perhaps even submit a PR).

However, from a practical standpoint, the most common scenario is that when there is a component with several paints (such as shape with background and border, and perhaps some text), then it is useful to be able to fade out (or fade in) that component as one whole, which involves maintaining ratios between opacities of individual paints.

ufrshubham replied:
Added a `_MultiPaintOpacityProvider` which allows creation of a provider based on a list of paint ids

st-pasha commented on packages/flame/lib/src/effects/provider_interfaces.dart:
```diff
@@ -50,3 +50,9 @@ abstract class SizeProvider {
   set size(Vector2 value);
 }
+
+/// Interface for a component that can be affected by opacity effects.
```
It would be good to document that setting the opacity to some `x` would not necessarily produce the same `x` when reading it back -- the implementation is allowed to store the opacity as an integer `alpha` which would incur some rounding errors.

ufrshubham replied:
This sounds more like implementation details. No need for the interface to make any comments on that.

st-pasha replied:
Indeed, it's an implementation detail. But it is a non-obvious detail. So the interface should tell the implementers that this is something which is allowed; and to the consumers that this is something that they could expect when using this API.

ufrshubham replied:
Done. Added the requested comments.

st-pasha commented on packages/flame/lib/src/components/mixins/has_paint.dart:
```diff
@@ -127,3 +148,51 @@ class _ProxyOpacityProvider<T extends Object> implements OpacityProvider {
+
+  @override
+  double get opacity {
+    var maxOpacity = 0.0;
```
```suggestion
    var maxOpacity = target.getOpacity();
```

ufrshubham replied:
That results in incorrect output. `getOpacity()` returns the opacity of `paint` which is not part of the `_paints` map. So if users provide a list of `paintIds`, `maxOpacity` will first get initialised with the default value of 1 from `paint`s opacity. As a result the result of `max` inside the loop will always be 1.

st-pasha commented on packages/flame/lib/src/components/mixins/has_paint.dart:
```diff
@@ -127,3 +148,51 @@ class _ProxyOpacityProvider<T extends Object> implements OpacityProvider {
+        : null;
+  }
+
+  final List<T>? paintIds;
```
```suggestion
  final List<T?>? paintIds;
```
since `null` is also a valid paint id

st-pasha commented on packages/flame/lib/src/components/mixins/has_paint.dart:
```diff
@@ -110,9 +112,28 @@ mixin HasPaint<T extends Object> on Component implements OpacityProvider {
+  /// Note: Each call results in a new [OpacityProvider] and hence the cached
+  /// opacity ratios are calculated using opacities when this method was called.
+  OpacityProvider opacityProviderOfList({List<T>? paintIds}) {
+    return _MultiPaintOpacityProvider(paintIds, this);
```
well `paintIds` is null, this should default to using all paints, as the description implies.

spydon commented on packages/flame/lib/src/components/mixins/has_paint.dart:
```diff
@@ -97,4 +100,91 @@ mixin HasPaint<T extends Object> on Component {
+  set opacity(double value) => target.setOpacity(value, paintId: paintId);
+}
+
+class _MultiPaintOpacityProvider<T extends Object> implements OpacityProvider {
```
Where/how is this used?

ufrshubham replied:
`opacityProviderOfList()` uses this. Using that method, users can specify a list of paintIds on which the provider should operate.

ufrshubham commented:
> It would be good with a few lines in the docs describing how to use this with the effect. :)

Updated the docs and also modified the `Flower` class so that it can be used to demonstrate `OpacityProvider` in one of the live examples.

ufrshubham commented on doc/flame/examples/lib/flower.dart:
```diff
@@ -50,7 +53,10 @@ class Flower extends PositionComponent with TapCallbacks {
-      canvas.drawPath(_paths[i], _paints[i]);
+      canvas.drawPath(
+        _paths[i],
+        getPaint(FlowerPaint.values.elementAt(i)),
```
> This will be a good usage of the layers in #2072 later!

Assuming this gets merged soon 😅

mattmyne replied:
I'm a bit concerned how well all this is going to play with paintLayers. It seems only the main paint and paintIds are taken into account (understandably given paintLayers isn't merged yet!) but hopefully this will be soon (maybe?). I haven't looked at this PR till now myself but - what do you think?

erickzanardo commented on doc/flame/effects.md:
```diff
@@ -376,6 +378,9 @@ uses multiple paints, the effect can target any individual color using the `pain
 final effect = OpacityEffect.to(
   0.2,
   EffectController(duration: 0.75),
+  target: component.opacityProviderOfList(
```
Should we mention and give an example of passing a target and not passing a target since this is optional?

ufrshubham replied:
Added this:

![OpacityEffectTargetDemo](https://user-images.githubusercontent.com/33748002/197253826-8571d2ef-b51a-4002-8611-3e333668205b.gif)

spydon commented:
Great job as always @ufrshubham! :)
